src/core/public/overlays/modal/modal_service.tsx

```diff
@@ -55,57 +56,79 @@ export class ModalRef implements OverlayRef {
+   * `close()` on the returned OverlayRef to close the modal.
+   *
+   * @param mount {@link MountPoint} - Mounts the children inside the modal
+   * @param options {@link OverlayModalOpenOptions} - options for the modal
+   * @return {@link OverlayRef} A reference to the opened modal.
+   */
+  open(mount: MountPoint, options?: OverlayModalOpenOptions): OverlayRef;
+}
+
+/**
+ * @public
+ */
+export interface OverlayModalOpenOptions {
+  className?: string;
+  closeButtonAriaLabel?: string;
+  'data-test-subj'?: string;
+}
+
+interface StartDeps {
+  i18n: I18nStart;
+  targetDomElement: Element;
+}
+
 /** @internal */
 export class ModalService {
   private activeModal: ModalRef | null = null;
+  private targetDomElement: Element | null = null;
 
-  constructor(private readonly targetDomElement: Element) {}
+  public start({ i18n, targetDomElement }: StartDeps): OverlayModalStart {
+    this.targetDomElement = targetDomElement;
 
-  /**
-   * Opens a flyout panel with the given component inside. You can use
-   * `close()` on the returned FlyoutRef to close the flyout.
-   *
-   * @param flyoutChildren - Mounts the children inside a flyout panel
-   * @return {FlyoutRef} A reference to the opened flyout panel.
-   */
-  public openModal = (
-    i18n: I18nStart,
-    modalChildren: React.ReactNode,
-    modalProps: {
-      closeButtonAriaLabel?: string;
-      'data-test-subj'?: string;
-    } = {}
-  ): ModalRef => {
-    // If there is an active flyout session close it before opening a new one.
-    if (this.activeModal) {
-      this.activeModal.close();
-      this.cleanupDom();
-    }
+    return {
+      open: (mount: MountPoint, options: OverlayModalOpenOptions = {}): OverlayRef => {
+        // If there is an active flyout session close it before opening a new one.
+        if (this.activeModal) {
+          this.activeModal.close();
+          this.cleanupDom();
+        }
 
-    const modal = new ModalRef();
+        const modal = new ModalRef();
 
-    // If a modal gets closed through it's ModalRef, remove it from the dom
-    modal.onClose.then(() => {
-      if (this.activeModal === modal) {
-        this.cleanupDom();
-      }
-    });
+        // If a modal gets closed through it's ModalRef, remove it from the dom
+        modal.onClose.then(() => {
+          if (this.activeModal === modal) {
+            this.cleanupDom();
+          }
+        });
 
-    this.activeModal = modal;
+        this.activeModal = modal;
 
-    render(
-      <EuiOverlayMask>
-        <i18n.Context>
-          <EuiModal {...modalProps} onClose={() => modal.close()}>
-            {modalChildren}
-          </EuiModal>
-        </i18n.Context>
-      </EuiOverlayMask>,
-      this.targetDomElement
-    );
+        render(
+          <EuiOverlayMask>
+            <i18n.Context>
+              <EuiModal {...options} onClose={() => modal.close()}>
+                <MountWrapper mount={mount} />
```

Comment:
@pgayvallet I am just a little confused as to why this wrapper needs to be **within** the EuiModal but wrapping the Modal contents. The EUI modal has a very specific component structure with styles that depend on that component structure. If we introduce a new element in between this could cause the layout to fail.

I do see you're trying to mitigate this with the styles you've applied to `.kbnMountWrapper`, however, this can become hard to manage if we ever make updates to the layout within EUI. Who will take ownership over making sure this doesn't break in future updates of EUI?

Basically, is the **any** way to either move this needed wrapping div to wrap the **entire** EuiModal, flyout, etc... Or can `MountWrapper` **not** spit out a `div`?

Reply:
To my knowledge, I'm afraid the answer for both is no...

I know the modal styles and tree structure is very precise and specific. However, EUI is react, and the goal of the kibana core API is to be framework agnostic. For this, we are now manipulating mount function instead of react element when calling the API and converting back into the react world using this `MountWrapper` which is using a reference to re-inject 'vanilla' dom inside the react tree. This can *only* be done by injecting a new element inside the structure (well, this can only be done by *having* a react element reference to inject into).

But now that I think about it, do you think that using
```typescript
export const EuiModalWrapper: React.FunctionComponent<{ mount: MountPoint }> = ({ mount, ...rest }) => {
  const element = useRef(null);
  useEffect(() => mount(element.current!), [mount]);
  return <EuiModal ref={element} {...rest} />;
};
```
instead of current `MountWrapper` is something that should work?

Reply:
Hum, just tested it, the answer is no, cannot put a ref on a react component, only on pure dom.

So

> Can MountWrapper not spit out a div

Rendering back to react requires a valid DOM ref, and that cannot be obtained without an element. So, no, not without modifications upstream to allow passing out new mountPoint to EUI instead of only react children, which is probably not an option. 

> is there any way to either move this needed wrapping div to wrap the entire EuiModal

Not without modifying every calls to the service, which loose all the interest of the core overlays services

> Who will take ownership over making sure this doesn't break in future updates of EUI?

I guess that will be the platform team responsability. 'good' new is that currently the functional tests detected the regression that forced me to add these styles, so we are no totally blind, at least in term of layout (not cosmetic)

CC @joshdover

Reply:
I see that (at least of flyouts), the `EuiFlyout` component keep a ref of itself in the `flyout` property.
So
```
export const EuiFlyoutMountWrapper: React.FunctionComponent<{
  mount: MountPoint;
  options: OverlayFlyoutOpenOptions;
}> = ({ mount, options }) => {
  const element = useRef(null);
  useEffect(() => mount(element.current!.flyout), [mount]);
  return <EuiFlyout ref={element} {...options} />;
};
```
This works, however as we alter a live react tree instead of properly injecting, we got a warning
`Warning: render(...): Replacing React-rendered children with a new root component. If you intended to update the children of this node, you should instead have the existing children update their state and render the new components instead of calling ReactDOM.render.`

Which is probably worse than introducing a new element in the tree.

Reply:
Thank you for attempting those other solutions. I'm going to pass the review to @chandlerprall instead for more technical knowledge.

Reply:
It seems to me that issue stems from combining multiple data sources into one output, i.e. the modal's header, body, and footer is being jammed into one location. Would it be possible instead to expand the overlay API into receiving those 3 data sets as distinct values and then independently mount them into the modal within the intended locations/components?

Reply:
@chandlerprall I'm not sure that really changes the issue, if we split the content info head/body/footer, the root problem remains the same: we will still need to get back to the react tree as some point, and therefor mounting back to (and injecting) 3 distinct elements instead of one. 

That's also major change for any modal/flyout with behaviour shared between these 3 sections, at the modal content and behaviour will need to be split into 3 distinct component parts before calling the API.

Reply:
One option would be to allow some EUI components to have an api to ease that kind of `non-react` usages. Either accept a `render` parameter similar to our `MountPoint`, of any other way to safely inject DOM to their react tree.

```jsx
export const EuiFlyoutMountWrapper: React.FunctionComponent<{
  mount: MountPoint;
  options: OverlayFlyoutOpenOptions;
}> = ({ mount, options }) => {
  const element = useRef(null);
  useEffect(() => mount(element.current!.flyout), [mount]);
  return <EuiFlyout ref={element} {...options} />;
};
```
'almost' works, and would (probably) if the inner reference used was of an empty content container instead of a reference were both `children` and other content are rendered at the same level. However that still feel hackier than a proper wrapper or real API to allow agnostic content to be rendered in EUI.

But once again, EUI is made in React and I'm not sure beeing 'pseudo-agnostic' is something you are looking for. However, it is a requirement for the core UI apis to be. That's why the 'wrapper with some layout styles' felt like an acceptable compromise, even if I do agree, that can be fragile in case of EUI style update, and any cleaner solution would be great. But having to adapt (heavily) every calls to our `modal` and `flyout` APIs is really something we would like to avoid if possible.

Reply:
>  if we split the content info head/body/footer, the root problem remains the same

That problem remains the same, but it resolves the issue of messing with EuiModal's expected DOM tree. EUI has a handful of components with this limitation/expectation, and unfortunately TypeScript [currently doesn't yet allow enforcing](https://github.com/microsoft/TypeScript/issues/21699), but we do [have plans](https://github.com/elastic/eui/issues/1890) to enforce it when TS resolves their issue (obviously a `@ts-ignore` gets around that).

> That's also major change for any modal/flyout with behaviour shared between these 3 sections

> But having to adapt (heavily) every calls to our modal and flyout APIs is really something we would like to avoid if possible.

Totally understand! As you said, _I guess that will be the platform team responsability_, which I think is enough to continue with this PR's approach. I'm willing to talk through other ideas/strategies to make these connections more seamless in the future, as well.

Reply:
> That problem remains the same, but it resolves the issue of messing with EuiModal's expected DOM tree

That's a very valid point.

To my knowledge, we only have a few core apis that actually rely on EUI (and therefor react). So my feeling is like being pragmatic here is probably the best thing (well, as long as pragmatism is acceptable for everyone). 

I know how making a framework opinionated library 'agnostic-friendly' can be hard and tedious to maintain. That's why as long as we only need to 'trick' with custom style on wrapper elements for `modal` `flyout` and maybe `toasts` in https://github.com/elastic/kibana/issues/36425, I feel like this is the best-compromise approach, and I'm not too afraid about maintenance (which is basically making sure that our wrapper mimics's it's parent's layout)

If at some point we extends our 'core UI' APIs with some more complex EUI usages however, I'm pretty convinced we will need a more seamless integration and that we will need to find a better integration.